scalding-core/src/main/scala/com/twitter/scalding/CascadingMode.scala

```
@@ -38,6 +40,28 @@ trait CascadingMode extends Mode {
   def fileExists(filename: String): Boolean
   /** Create a new FlowConnector for this cascading planner */
   def newFlowConnector(props: Config): FlowConnector
+
+  /**
+   * Make sure we are using our `ScaldingHfs` for `Hfs` taps.
+   */
+  protected def checkTap(tap: Tap[_, _, _]): Unit = {
+    tap match {
+      case hfs: Hfs =>
+        assert(
```

should we throw here? this seems like it might be very disruptive... 

Maybe a log WARN?

---

WARN will be invisible because we log a lot of stuff, and in this case, I think the exception is good since `toIterator` will not work otherwise.

---

won't it only not work for the elephantbird cases?

---

and, I mean, if it doesn't work, isn't it fine to let that exception kill things?

---

I expecting that parquet will be broken as well (they use the same pattern to read the first record in the constructor), but in general any record reader which read the first record in the constructor, and return them via `createValue` and `createKey`.

---

> and, I mean, if it doesn't work, isn't it fine to let that exception kill things?

it will be the less obvious exception, in some circumstances it will be hidden by cascading (because of their  `TupleEntryChainIterator` https://github.com/Cascading/cascading/blob/2.6.1/cascading-core/src/main/java/cascading/tuple/TupleEntryChainIterator.java#L77 also broken and hide exception by using `hashNext` in `next` method).

---

I'm just worried that this means that toIterator is broken for every single user that made their own Source.

Like, we may have seen this error once or twice at Stripe, but didn't diagnose it, but we haven't seen problems with toIterableExecution on non-source types.

---

> I'm just worried that this means that toIterator is broken for every single user that made their own Source.

for every custome Source which is not extending any standard one (what I think very rare).

> Like, we may have seen this error once or twice at Stripe, but didn't diagnose it, but we haven't seen problems with toIterableExecution on non-source types.

because usually `toIterableExecution` is force on disk snapshot of data in single file. 

what do you think to cover this assert by configuration key?

---

Put those asserts under configuration key, so by default, they are disabled.